### PR TITLE
CNF-13894: Hardware provisioning POC

### PR DESCRIPTION
**Pull request by tliu2021**
Summary
This update introduces changes to the IMS operator to enhance the handling of cluster requests. Specifically, it:
- Creates NodePool resources upon receiving a cluster request.
- Monitors the status of these NodePool resources as updated by the hardware manager plugin.
- Retrieves BMC information from node objects once the NodePool is provisioned with a list of node names.
- Populates the BMC information into the cluster instance and subsequently creates the cluster instance object.

Detailed Changes
1. NodePool Resource Creation and Monitoring:
-  On cluster request, the IMS operator now creates NodePool resources.
- The status of these NodePool resources is monitored, reflecting updates from the hardware manager plugin.

2. BMC Information Handling:
- Upon provisioning of the NodePool, the operator retrieves BMC information from the node objects.
- It then copies the BMC secret from the plugin namespace to the cluster namespace.
- This BMC information is then populated into the cluster instance, leading to the creation of the cluster instance object.

3. Interim BMC Secrets Handling:
- The existing code for creating BMC secrets is retained temporarily.
- These secrets will continue to be created until the hardware manager plugin provides the necessary information.

4. Node Resource Definition:
- The node resource definition is included in this update for now.
- This definition will be removed once it is provided by the hardware manager plugin.

Future Work
Unit Tests:
- Unit tests for these changes will be added in a separate update.

### CODE REVIEW

File: api/v1alpha1/node.go [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+	NodePool  string `json:"nodePool"`
+	GroupName string `json:"groupName"`
+	HwProfile string `json:"hwProfile"`
+}
+
+type BMC struct {
+	// Address contains the URL for accessing the BMC over the network.
+	Address string `json:"address,omitempty"`
+
+	// CredentiasName is a reference to a secret containing the credentials. That secret
```
Comment (donpenney) [inline api/v1alpha1/node.go:32]:
```suggestion
	// CredentialsName is a reference to a secret containing the credentials. That secret
```

File: api/v1alpha1/node.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+	// should contain the keys `username` and `password`.
+	CredentialsName string `json:"credentialsName,omitempty"`
+}
+
+// NodePoolStatus describes the observed state of a request to allocate and prepare
+// a node that will eventually be part of a deployment manager.
+type NodeStatus struct {
+	BMC *BMC `json:"bmc,omitempty"`
+	// Conditions represent the observations of the current state of the NodePool. Possible
+	// values of the condition type are `Provisioned`, `Unprovisioned``, `Updating` and `Failed`.
```
Comment (donpenney) [inline api/v1alpha1/node.go:42]:
```suggestion
	// values of the condition type are `Provisioned`, `Unprovisioned`, `Updating` and `Failed`.
```

File: api/v1alpha1/node_pools.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+}
+
+// NodePoolStatus describes the observed state of a request to allocate and prepare
+// a node that will eventually be part of a deployment manager.
+type NodePoolStatus struct {
+	// Properties represents the node properties in the pool
+	Properties Properties `json:"properties,omitempty"`
+
+	// Conditions represent the observations of the current state of the NodePool. Possible
+	// values of the condition type are `Provisioned`, `Unprovisioned``, `Updating` and `Failed`.
```
Comment (donpenney) [inline api/v1alpha1/node_pools.go:57]:
```suggestion
	// values of the condition type are `Provisioned`, `Unprovisioned`, `Updating` and `Failed`.
```

File: internal/controllers/clusterrequest_controller.go [hunk 4]
```diff
@@ -749,34 +961,59 @@ func (r *ClusterRequestReconciler) finalizeClusterRequest(
...
+					if err != nil {
+						return err
+					}
+				}
 			}
 
-			if exists {
-				err := r.Client.Delete(ctx, clusterInstance)
+			if clusterRequest.Status.HardwareProvisioningStatus != nil {
+				// delete the node pool created by this request if it exits
```
Comment (donpenney) [inline internal/controllers/clusterrequest_controller.go:997]:
```suggestion
				// delete the node pool created by this request if it exists
```

File: internal/controllers/clusterrequest_controller.go [hunk 5]
```diff
@@ -833,6 +1070,200 @@ func (r *ClusterRequestReconciler) handleFinalizer(
 	return ctrl.Result{}, false, nil
 }
 
+func (t *clusterRequestReconcilerTask) waitForHardwareData(ctx context.Context,
+	clusterInstance *unstructured.Unstructured, nodePool *oranv1alpha1.NodePool) bool {
+
+	// Get the generated NodePool and its status.
+	exits, err := utils.DoesK8SResourceExist(ctx, t.client, nodePool.GetName(),
```
Comment (donpenney) [inline internal/controllers/clusterrequest_controller.go:1077]:
```suggestion
	exists, err := utils.DoesK8SResourceExist(ctx, t.client, nodePool.GetName(),
```

File: internal/controllers/clusterrequest_controller.go [hunk 6]
```diff
@@ -833,6 +1070,200 @@ func (r *ClusterRequestReconciler) handleFinalizer(
...
+		nodePool.GetNamespace(), nodePool)
+
+	if err != nil || !exits {
```
Comment (donpenney) [inline internal/controllers/clusterrequest_controller.go:1080]:
```suggestion
	if err != nil || !exists {
```

File: internal/controllers/utils/constants.go [hunk 7]
```diff
@@ -92,3 +92,17 @@ const (
 	ClusterZtpDone    = "ZTP Done"
 	ClusterZtpNotDone = "ZTP Not Done"
 )
+
+// Hardware Privisioning status
+const (
+	HardwareProvisioning         = "In progress"
+	HardwareProvisioningComplted = "Completed"
```
Comment (donpenney) [inline internal/controllers/utils/constants.go:99]:
```suggestion
	HardwareProvisioningCompleted = "Completed"
```

File: internal/controllers/utils/constants.go [hunk 8]
```diff
@@ -92,3 +92,17 @@ const (
...
+	HardwareProvisioningFailed   = "Failed"
+	HardwareProvisioningUnkown   = "Unkown"
```
Comment (donpenney) [inline internal/controllers/utils/constants.go:101]:
```suggestion
	HardwareProvisioningUnknown   = "Unknown"
```

File: internal/controllers/clusterrequest_controller.go [hunk 9]
```diff
@@ -833,6 +1070,200 @@ func (r *ClusterRequestReconciler) handleFinalizer(
...
+	}
+
+	t.object.Status.HardwareProvisioningStatus.ClusterName = nodePool.GetName()
+	t.object.Status.HardwareProvisioningStatus.Namespace = nodePool.GetNamespace()
+
+	if len(nodePool.Status.Conditions) > 0 {
+		provisionedCondition := meta.FindStatusCondition(
+			nodePool.Status.Conditions, "Provisioned")
+		if provisionedCondition != nil && provisionedCondition.Status == metav1.ConditionTrue {
+			t.object.Status.HardwareProvisioningStatus.HardwareStatus = utils.HardwareProvisioningComplted
```
Comment (donpenney) [inline internal/controllers/clusterrequest_controller.go:1221]:
```suggestion
			t.object.Status.HardwareProvisioningStatus.HardwareStatus = utils.HardwareProvisioningCompleted
```

File: internal/controllers/clusterrequest_controller.go [hunk 10]
```diff
@@ -833,6 +1070,200 @@ func (r *ClusterRequestReconciler) handleFinalizer(
...
+				nodePool.Status.Conditions, "Provisioning")
+			if provisioningCondition != nil && provisioningCondition.Status == metav1.ConditionTrue {
+				t.object.Status.HardwareProvisioningStatus.HardwareStatus = utils.HardwareProvisioning
+			} else {
+				failedCondition := meta.FindStatusCondition(
+					nodePool.Status.Conditions, "Failed")
+				if failedCondition != nil && failedCondition.Status == metav1.ConditionTrue {
+					t.object.Status.HardwareProvisioningStatus.HardwareStatus = utils.HardwareProvisioningFailed
+				} else {
+					t.object.Status.HardwareProvisioningStatus.HardwareStatus = utils.HardwareProvisioningUnkown
```
Comment (donpenney) [inline internal/controllers/clusterrequest_controller.go:1233]:
```suggestion
					t.object.Status.HardwareProvisioningStatus.HardwareStatus = utils.HardwareProvisioningUnknown
```

File: internal/controllers/utils/constants.go [hunk 11]
```diff
@@ -92,3 +92,17 @@ const (
 	ClusterZtpDone    = "ZTP Done"
 	ClusterZtpNotDone = "ZTP Not Done"
 )
+
+// Hardware Privisioning status
```
Comment (donpenney) [inline internal/controllers/utils/constants.go:96]:
```suggestion
// Hardware Provisioning status
```

File: api/v1alpha1/node.go (hunk 1 above)
Comment (tliu2021) [reply api/v1alpha1/node.go]: Done

File: api/v1alpha1/node.go (hunk 2 above)
Comment (tliu2021) [reply api/v1alpha1/node.go]: Done

File: api/v1alpha1/node_pools.go (hunk 3 above)
Comment (tliu2021) [reply api/v1alpha1/node_pools.go]: Done

File: internal/controllers/clusterrequest_controller.go (hunk 4 above)
Comment (tliu2021) [reply internal/controllers/clusterrequest_controller.go]: Done

File: internal/controllers/clusterrequest_controller.go (hunk 5 above)
Comment (tliu2021) [reply internal/controllers/clusterrequest_controller.go]: Done

File: internal/controllers/clusterrequest_controller.go (hunk 6 above)
Comment (tliu2021) [reply internal/controllers/clusterrequest_controller.go]: Done

File: internal/controllers/utils/constants.go (hunk 7 above)
Comment (tliu2021) [reply internal/controllers/utils/constants.go]: Done

File: internal/controllers/utils/constants.go (hunk 8 above)
Comment (tliu2021) [reply internal/controllers/utils/constants.go]: Done

File: internal/controllers/clusterrequest_controller.go (hunk 9 above)
Comment (tliu2021) [reply internal/controllers/clusterrequest_controller.go]: Done

File: internal/controllers/clusterrequest_controller.go (hunk 10 above)
Comment (tliu2021) [reply internal/controllers/clusterrequest_controller.go]: Done

File: internal/controllers/utils/constants.go (hunk 11 above)
Comment (tliu2021) [reply internal/controllers/utils/constants.go]: Done

File: internal/controllers/utils/constants.go [hunk 12]
```diff
@@ -92,3 +92,17 @@ const (
 	ClusterZtpDone    = "ZTP Done"
 	ClusterZtpNotDone = "ZTP Not Done"
 )
+
+// Hardware Provisioning status
+const (
+	HardwareProvisioning          = "In progress"
```
Comment (pixelsoccupied) [inline internal/controllers/utils/constants.go:98]: Should it be `HardwareProvisioningInProgress`?

File: internal/controllers/clusterrequest_controller.go [hunk 13]
```diff
@@ -833,6 +1070,200 @@ func (r *ClusterRequestReconciler) handleFinalizer(
 	return ctrl.Result{}, false, nil
 }
 
+func (t *clusterRequestReconcilerTask) waitForHardwareData(ctx context.Context,
```
Comment (pixelsoccupied) [inline internal/controllers/clusterrequest_controller.go:1073]: Could you add a comment describing what this function is doing? Slightly hard to follow...but it's not just waiting (deciding if ready) but final goal is to update NodePool yeah? Would nice split into two e.g `IsPluginManagerReady` and `UpdateNodePool` 

Also shouldnt we bubble up these errors and show it in status?

File: api/v1alpha1/node_pools.go [hunk 14]
```diff
@@ -0,0 +1,88 @@
...
+	//
+	// +kubebuilder:validation:Required
+	CloudID string `json:"cloudID"`
+
+	// Location is the geographical location of the requested node.
+	//
+	// +kubebuilder:validation:Required
+	Location string `json:"location"`
+
+	Site string `json:"site"`
```
Comment (pixelsoccupied) [inline api/v1alpha1/node_pools.go:35]: Given that these two vars are related and talking about GeoLocation (and also used in ClusterSpec)....can we simply put in a `type GeoLocation struct` and share them between the two specs instead of them being at the top level?

File: api/v1alpha1/clusterrequest_types.go [hunk 15]
```diff
@@ -29,6 +29,10 @@ type ClusterRequestSpec struct {
 	// INSERT ADDITIONAL SPEC FIELDS - desired state of cluster
 	// Important: Run "make" to regenerate code after modifying this file
 
+	Location string `json:"location"`
+
+	Site string `json:"site"`
```
[the remark this answers is not in the document]
Comment (tliu2021) [reply api/v1alpha1/clusterrequest_types.go]: Yes, done.

File: internal/controllers/clusterrequest_controller.go [hunk 16]
```diff
@@ -152,8 +153,34 @@ func (t *clusterRequestReconcilerTask) run(ctx context.Context) (nextReconcile c
...
+	// ### HARDWARE TEMPLATE RENDERING ###
+	renderedNodePool, renderErr := t.renderHardwareTemplate(ctx, renderedClusterInstance)
+	if renderErr != nil {
+		return ctrl.Result{}, nil
+	}
+
+	// ### CREATE/UPDATE NODE POOL CR ###
+	createErr := t.createNodePoolResources(ctx, renderedNodePool)
+	if createErr != nil {
+		return ctrl.Result{}, nil
```
[the remark this answers is not in the document]
Comment (tliu2021) [reply internal/controllers/clusterrequest_controller.go:165]: I reviewed the rest of the code and noticed that we don’t immediately requeue for a retry on errors. To maintain consistency, I applied the same approach in the new code.

File: internal/controllers/utils/constants.go (hunk 12 above)
Comment (tliu2021) [reply internal/controllers/utils/constants.go:98]: Provisioning means “In process” :-)

File: internal/controllers/clusterrequest_controller.go (hunk 13 above)
Comment (tliu2021) [reply internal/controllers/clusterrequest_controller.go]: A comment is added and it is split into two functions.   It is to be determined  whether to display the internal errors in the status.

File: api/v1alpha1/node_pools.go (hunk 14 above)
Comment (tliu2021) [reply api/v1alpha1/node_pools.go]: Added a LocationSpec struct.  Since 'Site' is mapped to a 'location' per Dell,  I made location optional.

File: internal/controllers/utils/constants.go (hunk 12 above)
Comment (pixelsoccupied) [reply internal/controllers/utils/constants.go:98]: True! It was just for consistency i guess (name + what it's doing...pattern)

I guess the "best" way to deal with this would be have typed status 

```go
type HardwareProvisioningStatus string

InProgress  HardwareProvisioningStatus = "InProgress"
Failed      HardwareProvisioningStatus = "Failed"

etc etc 
```

But don't have any strong preference 

Btw totally missed it but should be "InProgress" (conditions types are expected to be PascalCase)

File: internal/controllers/clusterrequest_controller.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (pixelsoccupied) [reply internal/controllers/clusterrequest_controller.go:165]: Yeah seems kind of random to just stop it. Hope we can discuss this later. But good for now.

[the remark this answers is not in the document]
Comment (tliu2021) [reply internal/controllers/clusterrequest_controller.go:165]: Yes, I agree. We should schedule an update to revisit all error conditions.

File: internal/controllers/utils/constants.go (hunk 12 above)
Comment (tliu2021) [reply internal/controllers/utils/constants.go]: Changed.

File: api/v1alpha1/node_pools.go [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+type LocationSpec struct {
+	Location string `json:"location,omitempty"`
+	Site     string `json:"site"`
```
Comment (bartwensley) [inline api/v1alpha1/node_pools.go:23]: Looking at the "O-RAN O2ims Interface Specification 5.0" I only see mention of the location (no mention of site). Lets just stick with "location" for now unless someone else has seen "site" in the O-RAN documents somewhere. I will update this in the design as well.

Comment (tliu2021) [reply api/v1alpha1/node_pools.go:23]: Per Dell, the compute resource pool(s) are mapped to a site (while a site is created on a location). See reference: https://docs.google.com/document/d/19lPxbaQkb-V8JNL3JJXCMEqivNqGmucOF48IAu7FOCk/edit

**Comment by pixelsoccupied** [comment]
/lgtm

Good stuff!
